GenericList/list.h

```diff
@@ -0,0 +1,260 @@
+
+        std::move(data + index + 1, data + count, data + index);
+        data[--count].~T();
+    }
+
+
+    //Returns true if a relevant element exists, and removes it.
+    size_t Remove(const T& val) {
+        //Lazy, simpler way - when changing behaviour, just change it in RemoveIf:
+        return RemoveIf([&](const auto& e) { return e == val; } );
+    }
+    
+    template <typename Predicate>
+    size_t RemoveIf(Predicate&& pred) {
+        //idea: double pointers, picker points to next element to check,
+        //                       placer points to next available space
+
+        auto placer = FindIf(pred);
+        if (placer == nullptr) return 0;
+
+        for (auto picker = placer + 1; picker < end(); ++picker) {
+            if (!pred(*picker)) {
+                *placer = std::move(*picker); //*picker now contains irrelevant instantiated data
+                ++placer;
+            }
+        }
+
+        //if constexpr is evaluated at compile time - eg: List<int> won't have the following code when compiled
+        if constexpr (!std::is_trivially_destructible<T>::value) { //those that are don't have non-empty destructors to call
+            std::cout << "Not trivially destructible!" << "\n";
+            for (auto k = placer; k < end(); ++k) {
+                k->~T(); //destructs meaningless data from matched indexes moved to end
+            }
+        }
+
+        const auto removed = end() - placer;
+        count -= removed;
+
+        return removed;
+    }
+
+    const T& operator[](size_t index) const { return data[index]; } //read-only
+    T& operator[](size_t index) { return data[index]; } //read+(later)write
+    const T& Get(size_t index) const {
+        if (index >= count || index < 0) throw std::out_of_range(std::string("Cannot get element at out_of_range index: ") + std::to_string(index) + std::string(")"));
+
+        return data[index];
+    }
+    T& Get(size_t index) {
+        if (index >= count || index < 0) throw std::out_of_range(std::string("Cannot get element at out_of_range index: ") + std::to_string(index) + std::string(")"));
+
+        return data[index];
+    }
+
+
+    //can iterate through list without using iterators if internal data is contiguous
+    T* begin() { return data; }
+    const T* begin() const { return data; } //non-copy version for read-only
+    const T* cbegin() const { return data; } //same as above, allows for direct call to cbegin for user who knows they want constant iterators
+    
+    T* end() { return data + count; }
+    const T* end() const { return data + count; }
+    const T* cend() const { return data + count; }
+
+
+
+private:
+    T* data;
+    size_t capacity;
+    size_t count;
+
+    Allocator dataAllocator; //not actually necessary to maintain same allocator throughout program as allocators can allocate/deallocate any data
+
+    static T* CreateDeepCopy(T* data, size_t data_size, size_t copy_size, Allocator alloc) {
```

Comment:
Don't copy the allocator, pass it by reference. You could also just pass the whole class instance:
```suggestion
    static T* CreateDeepCopy(List const&, Allocator& alloc_to_use) {
```